Error Tracking Docs -- React Native (#31545)
DataDog/documentation: content/en/error_tracking/frontend/mobile/reactnative.md

code_lang_weight: 60
---

## Overview 

Enable React Native Crash Reporting and Error Tracking to get comprehensive crash reports and error trends. With this feature, you can access:

-   Aggregated React Native crash dashboards and attributes
-   Symbolicated React Native (JavaScript and native iOS or Android) crash reports
-   Trend analysis with React Native Error Tracking

**Note**: In order to symbolicate your stack traces, you must manually upload your source maps and native debug symbols into Datadog.

## Setup

If you have not set up the React Native SDK yet, follow the [in-app setup instructions][16] or see the [React Native setup documentation][17]. Then, follow the steps on this page to enable React Native Crash Reporting and Error Tracking.

 

### Step 1 - Specify application details in the UI

> 1. In Datadog, navigate to [**Error Tracking** > **Settings** > **Browser and Mobile** > **Add an Application**][1].

Reviewer: ```suggestion
1. In Datadog, navigate to [**Error Tracking** > **Settings** > **Browser and Mobile**][1] and click **+ New Application**.
```

Reviewer: There doesn't seem to be a way to link directly to the screen for adding an application. I've also suggested updating the URL on line `895` to be absolute, since relative links in the docs repo open other doc pages.

config.sessionSamplingRate = 80;
// Optional: Sample tracing integrations for network calls between your app and your backend (here, 80% of calls to your instrumented backend are linked from the RUM view to the APM view. Default = 20%).
// You need to specify the hosts of your backends to enable tracing with these backends
config.resourceTracingSamplingRate = 80;
config.firstPartyHosts = ['example.com']; // matches 'example.com' and subdomains like 'api.example.com'

//Wrap the content of your App component in a DatadogProvider component, passing it your configuration:

export default function App() {
    return (
        <DatadogProvider configuration={config}>
            <Navigation />
        </DatadogProvider>
    );
}
```
{{< /site-region >}}

<br>

### Step 3 - Add Crash Reporting

Update your initialization snippet to enable native JavaScript crash reporting:

```javascript
const config = new DdSdkReactNativeConfiguration(
    '<CLIENT_TOKEN>',
    '<ENVIRONMENT_NAME>',
    '<APPLICATION_ID>',
    true,
    true,
    true // enable JavaScript crash reporting
);
config.nativeCrashReportEnabled = true; // enable native crash reporting
```

> Your crash reports appear in [**Error Tracking**][1].

Reviewer: ```suggestion
Your crash reports appear in [**Error Tracking**][18].
```

[201]: https://github.com/DataDog/datadog-ci/blob/master/src/commands/react-native/README.md#inject-debug-id

{{% /collapse-content %}}

{{% collapse-content title="Passing options for your uploads" level="h5" %}}
##### Using the `datadog-sourcemaps.gradle` script

To specify a different service name, add the following code to your `android/app/build.gradle` file, before the `apply from: "../../node_modules/@datadog/mobile-react-native/datadog-sourcemaps.gradle"` line:

```groovy
project.ext.datadog = [
    serviceName: "com.my.custom.service"
]
```

##### Using the `datadog-ci react-native xcode` command

Options for the `datadog-ci react-native xcode` command are available on the [command documentation page][301].

##### Specifying a custom release version

Use the `DATADOG_RELEASE_VERSION` environment variable to specify a different release version for your source maps, starting from `@datadog/mobile-react-native@2.3.5` and `@datadog/datadog-ci@v2.37.0`.

When the SDK is initialized with a version suffix, you must manually override the release version in order for the source map and build versions to match.

> [301]: https://github.com/DataDog/datadog-ci/tree/master/src/commands/react-native#xcode

Reviewer: This link leads to a `404` error, and I'm not sure what to update it with since the folder structure of that repo has changed.

`.GRANTED`
: The React Native SDK starts collecting the data and sends it to Datadog.

`.NOTGRANTED`
: The React Native SDK does not collect any data. No logs, traces, or RUM events are sent to Datadog.

To change the tracking consent value and behavior after the React Native SDK is initialized, use the `Datadog.set(trackingConsent:)` API call.

For example, assuming the current tracking consent is `.PENDING`:

- If you change the `trackingConsent` value to `.GRANTED`, the React Native SDK sends all current and future data to Datadog;
- If you change the `trackingConsent` value to `.NOTGRANTED`, the React Native SDK wipes all current data and does not collect future data.

{{% /collapse-content %}}

{{% collapse-content title="Set sample session rates" level="h4" %}}

To control the data your application sends to Datadog, you can specify a sampling rate for sessions while [initializing the React Native SDK][101] as a percentage between 0 and 100. You can specify the rate with the `config.sessionSamplingRate` parameter.

> [101]: /real_user_monitoring/mobile_and_tv_monitoring/react_native/setup/reactnative/#initialize-the-library-with-application-context

Author: ```suggestion
[101]: /real_user_monitoring/application_monitoring/react_native/setup/reactnative?tab=rum#initialize-the-library-with-application-context
```

Add the following snippet during initialization in your Datadog configuration:

```javascript
configuration.trackBackgroundEvents = true;
```
{{% /collapse-content %}}

{{% collapse-content title="Send data when device is offline" level="h4" %}}

The React Native SDK ensures availability of data when your user device is offline. In cases of low-network areas, or when the device battery is too low, all events are first stored on the local device in batches. They are sent as soon as the network is available, and the battery is high enough to ensure the React Native SDK does not impact the end user's experience. If the network is not available with your application running in the foreground, or if an upload of data fails, the batch is kept until it can be sent successfully.

This means that even if users open your application while offline, no data is lost.

**Note**: The data on the disk is automatically deleted if it gets too old to ensure the React Native SDK does not use too much disk space.

{{% /collapse-content %}}

{{% collapse-content title="Data storage" level="h4" %}}
##### Android

> Before data is uploaded to Datadog, it is stored in cleartext in your application's cache directory. This cache folder is protected by [Android's Application Sandbox][11], meaning that on most devices this data can't be read by other applications. However, if the mobile device is rooted, or someone tampers with the Linux kernel, the stored data might become readable.

Reviewer: The link on this line is leading to a `404` error, and I'm not sure how it should be updated

##### iOS

> Before data is uploaded to Datadog, it is stored in cleartext in the cache directory (`Library/Caches`) of your [application sandbox][12], which can't be read by any other app installed on the device.

Reviewer: It seems like the link on this line is not opening to the expected page -- could you please confirm?

Author: Fixed this with link 20

{{% /collapse-content %}}

{{% collapse-content title="Development mode" level="h4" %}}
While in development mode, your application can submit extra events related to the React Native tooling, such as code transformation errors and requests to a local development server.

To prevent these events from showing in the dashboard, you can disable errors and resources tracking in dev mode using the `__DEV__` flag:

```js
const config = new DatadogProviderConfiguration(
	CLIENT_TOKEN,
	ENVIRONMENT,
	APPLICATION_ID,
	true,
	!__DEV__  /* trackResources will be false in DEV mode, true otherwise */,
	!__DEV__  /* trackErrors will be false in DEV mode, true otherwise */,
	trackingConsent
)
```
{{% /collapse-content %}}

{{% collapse-content title="New architecture support" level="h4" %}}
The [React Native new architecture][601] is supported by the React Native SDK in version `>=1.8.0`.

The minimum supported React Native version for the new architecture is `0.71`.

> [601]: https://source.android.com/security/app-sandbox

Reviewer: It looks like this might be the link intended for reference link `[12]` on line `453`, but not sure if this is the expected page for `React Native new architecture`. Could you please confirm?

  --bundle-output build/main.jsbundle \
  --sourcemap-output build/main.jsbundle.map

sourcemapsize=$(wc -c build/main.jsbundle.map | awk '{print $1}')
bundlesize=$(wc -c build/main.jsbundle | awk '{print $1}')
payloadsize=$(($sourcemapsize + $bundlesize))

echo "Size of source maps and bundle is $(($payloadsize / 1000000))MB"
```

If a `build` directory does not already exist, create it first by running `mkdir build`, then run the command above.

## Test your implementation

To verify your React Native Crash Reporting and Error Tracking configuration, you need to issue an error in your application and confirm that the error appears in Datadog.

To test your implementation:

1. Run your application on a simulator, emulator, or a real device. If you are running on iOS, ensure that the debugger is not attached. Otherwise, Xcode captures the crash before the Datadog SDK does.
2. Execute some code containing an error or crash. For example:

   ```javascript
   const throwError = () => {
    throw new Error("My Error")
   }
   ```

3. For obfuscated error reports that do not result in a crash, you can verify symbolication and deobfuscation in [**Error Tracking**][1].

> 4. For crashes, after the crash happens, restart your application and wait for the React Native SDK to upload the crash report in [**Error Tracking**][1].

Reviewer: As before, it sounds like these are expected to open to the main Error Tracking page in Datadog

To make sure your source maps are correctly sent and linked to your application, you can also generate crashes with the [`react-native-performance-limiter`][10] package.

> Install it with yarn or npm then re-install your pods:

Reviewer: ```suggestion
Install it with yarn or npm then reinstall your pods:
```

```shell
yarn add react-native-performance-limiter # or npm install react-native-performance-limiter
(cd ios && pod install)
```

Crash the JavaScript thread from your app:

```javascript
import { crashJavascriptThread } from 'react-native-performance-limiter';

const crashApp = () => {
    crashJavascriptThread('custom error message');
};
```

> Re-build your application for release to send the new source maps, trigger the crash and wait on the [Error Tracking][1] page for the error to appear.

Reviewer: ```suggestion
Rebuild your application for release to send the new source maps, trigger the crash and wait on the [Error Tracking][1] page for the error to appear.
```

            tasks["minify${variant.name.capitalize()}WithR8"].finalizedBy { tasks["uploadMapping${variant.name.capitalize()}"] }
        }
```

**Note**: Re-uploading a source map does not override the existing one if the version has not changed.

{{% /collapse-content %}}

## Troubleshooting

### Usage with `use_frameworks!`

If you have `use_frameworks!` enabled in your `Podfile`, running `pod install` after adding the SDK is likely to trigger an error like this one:

```shell
The 'Pods-MyApp' target has transitive dependencies that include statically linked binaries: (DatadogSDKBridge, DatadogSDKCrashReporting)
```

To prevent that error, edit your `Podfile` to install the React Native SDK pod as a static library:

```ruby
static_libraries = ['DatadogSDKReactNative']

# Turn pods with static dependencies into static libraries by overriding the static_framework? function to return true
pre_install do |installer|
  installer.pod_targets.each do |pod|
    if static_libraries.include?(pod.name)
      def pod.static_framework?;
        true
      end
      def pod.build_type;
        Pod::BuildType.static_library
      end
    end
  end
end
```

> **Note**: This solution comes from this [StackOverflow][8] post.

Reviewer: This isn't opening to a StackOverflow post, and I'm not seeing one in the list of reference links

## Further Reading

{{< partial name="whats-next/whats-next.html" >}}

[1]: /error_tracking/
[2]: https://microsoft.github.io/code-push/
[3]: /real_user_monitoring/mobile_and_tv_monitoring/react_native/setup/codepush
[4]: https://docs.expo.dev/
[5]: /real_user_monitoring/mobile_and_tv_monitoring/react_native/setup/expo
[6]: /account_management/api-app-keys/#api-keys
[7]: /account_management/api-app-keys/#client-tokens
[8]: https://github.com/DataDog/datadog-react-native-wizard
[9]: https://app.datadoghq.com/source-code/setup/rum
[10]: https://github.com/DataDog/react-native-performance-limite
[11]: https://github.com/DataDog/datadog-ci/tree/master/src/commands/react-native#xcode
[12]: /real_user_monitoring/mobile_and_tv_monitoring/ios/error_tracking/?tabs=cocoapods#symbolicate-crash-reports
[13]: https://reactnative.dev/docs/signed-apk-android#enabling-proguard-to-reduce-the-size-of-the-apk-optional
[14]: https://plugins.gradle.org/plugin/com.datadoghq.dd-sdk-android-gradle-plugin
[15]: https://github.com/DataDog/dd-sdk-android-gradle-plugin
[16]: https://app.datadoghq.com/rum/application/create

> [17]: /real_user_monitoring/mobile_and_tv_monitoring/react_native/

Reviewer: ```suggestion
[17]: /real_user_monitoring/application_monitoring/react_native/setup/

```

Reviewer: It looks like this is the link for React Native setup documentation -- could you please confirm?

> 

Reviewer: ```suggestion
[18]: https://app.datadoghq.com/error-tracking
```
